p4-16/psa/PSA.mdk

```diff
@@ -1497,6 +1499,59 @@ table.
 A PSA implementation must implement tables that have both a
 `psa_direct_counter` and `psa_direct_meter` property.
 
+### Table entry timeout notification {#sec-idle-timeout}
+
+PSA uses the `psa_idle_timeout` to enable a table implementation send
+notifications from the PSA device when a configurable time has passed since an
+entry was last matched. The property may take one of two values --
+`PSA_NO_TIMEOUT`, and `PSA_NOTIFY_CONTROL`. `PSA_NO_TIMEOUT` disables idle
+timeout support for the table. `PSA_NOTIFY_CONTROL` enables the notification. A
+PSA implementation will then generate an API for the control plane set
+time-to-live (TTL) values for table entries and if at any time during its
+lifetime, the table entry is not "hit" (&ie; not selected by any packet lookup)
+for a lapse of time greater or equal to its TTL, the device should generate a
+notification to the control plane.
+
+Example:
+```
+enum PSA_IdleTimeout_t {
+  PSA_NO_TIMEOUT,
+  PSA_NOTIFY_CONTROL
+}
+
+table t {
+  key = {};
+  actions = {};
+  psa_idle_timeout = PSA_NOTIFY_CONTROL;
+}
+```
+
+Restrictions on the TTL values and notifications:
+
+- It is likely that any hardware implementation will have a limited number of
+  bits to represent the values, and, since the values are programmed at
+  runtime, it is the responsability of the runtime (P4Runtime or other
+  controller software) to guarantee that the TTL values can be represented in
+  the device. This can be done by scaling the values to the number of bits
+  available on the platform, ensuring that the range of values between
+  different entries are representable. A PSA implementation should only enable
+  the programming of such tables, and return an error if the device does not
+  support the idle timeout at all.
+
+- If no value is programmed for a table entry, even though the table has
+  enabled the idle timeout property, the entry will not generate a
+  notification. The entry will participate in the aging process, with the
+  maximum timeout interval possible, i.e., its idle timeout counter will be set
+  to the maximum interval possible and the entry may be deleted when the
+  maximum interval elapses.
+
+- PSA does not require a timeout value for a default action entry. The reason
+  for not making this mandatory in the specification is that the default action
+  may not have an actual entry in the table and it is may be hit on the miss
+  table misses. Not sure what an TTL means for a miss path. The default action entry will not be aged out.
+
+- ActionSelectors and ActionProfiles will inherit the idle timeout property
```

Comment:
I think this is ambiguous. The idle timeout process does not depend on the table's implementation (direct or indirect). But that only includes the match entry part, not the member / group the entry is pointing to. Even if a member is only used by a single match entry and that entry expires, I don't think the member should be removed.

Reply:
If it gets too confusing/weird, we could consider explicitly saying that PSA does not (yet) specify how the idle timeout feature works with tables that have action selector or action profile implementations, and implementations need not support it in order to be PSA v1.1. compliant.  I believe the PSA spec already says that implementations need not support direct meters or direct counters on tables that have action profiles/selectors, and timeouts are similar to those things in some ways.

Reply:
I'm fine with making support optional. But I think it's fairly easy to have idle timeout semantics for match tables regarding of the implementation, by limiting the idle timeout feature to the match "part" of the entry.

Reply:
restricted and left for future work.